### PR TITLE
Devise

### PR DESCRIPTION
**Pull request by Ru-T**
## What Changed? 
* Added devise gem and config, including confirmable
* Override after sign up path in registrations controller
* Added login/logout links to the header
* Added integration tests for account creation and authentication

## Why? 
* So that users must log in in order to view any part of the site.

### CODE REVIEW

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -3,9 +3,21 @@
 <head>
   <title>RutiBookstore</title>
   <%= csrf_meta_tags %>
+  <%= javascript_include_tag "application", "data-turbolinks-track" => true %>
```
Comment (dacur) [inline app/views/layouts/application.html.erb]: it looks like you removed `turbolinks` (which you should!), so i'm not sure if this old reference to it here will cause you problems, but you can probably safely remove it

`<%= javascript_include_tag "application" %>`

File: config/application.rb [hunk 2]
```diff
@@ -22,5 +22,21 @@ class Application < Rails::Application
...
+    config.action_mailer.raise_delivery_errors = true
+    config.action_mailer.delivery_method = :smtp
+    config.action_mailer.smtp_settings = {
+      :port           => 587,
+      :address        => 'smtp.mailgun.org',
+      :user_name      => 'postmaster@sandbox85b185dc6fa44a7bbb05e2ce19beca69.mailgun.org',
+      :password       => '065168f69b0027f0cb9f05b917383d0a',
+      :domain         => 'sandbox85b185dc6fa44a7bbb05e2ce19beca69.mailgun.org',
+      :authentication => :plain
+    }
```
Comment (dacur) [inline config/application.rb]: you might consider moving this mailer code into its own config file and also using environment variables or something else to protect your mailer credentials

File: db/seeds.rb [hunk 3]
```diff
@@ -7,3 +7,14 @@
               description: Faker::Lorem.sentence
               )
 end
+
+
+User.create!(email: "rwajnberg@gmail.com",
+            password: "password"
+            )
+
```
Comment (dacur) [inline db/seeds.rb]: nice job creating your own login :+1:

File: features/step_definitions/account_creation_steps.rb [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+Then(/^I am told to check my email for a confirmation link$/) do
+  expect(page).to have_content("A message with a confirmation link has been sent to your email address. Please follow the link to activate your account.")
+end
+
+Then(/^I am sent a confirmation email$/) do
+  ActionMailer::Base.deliveries.last.body.match("You can confirm your account email through the link below:")
+end
+
+When(/^I visit the link in that email$/) do
+  ctoken = ActionMailer::Base.deliveries.last.body.match(/confirmation_token=\w*/)
```
Comment (dacur) [inline features/step_definitions/account_creation_steps.rb]: check out the `email-spec` gem.  i found it very helpful and easy to use.  (you don't have to change what you've done)

https://github.com/email-spec/email-spec

File: app/views/layouts/application.html.erb (hunk 1 above)
Comment (Ru-T) [inline app/views/layouts/application.html.erb]: You're right - accident that I added it back. :)

File: spec/factories/users.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+FactoryGirl.define do
+  factory :user do
+    email "newuser@test.com"
+    password "password"
```
Comment (dacur) [inline spec/factories/users.rb]: unless you need to know what these are, you could also use `Faker` here (not required)

File: spec/models/user_spec.rb [hunk 6]
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
```
Comment (dacur) [inline spec/models/user_spec.rb]: if you don't need this file, you can delete it.

File: features/authentication.feature [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    Then I am redirected to the book index page
+
+  Scenario: User enters invalid authentication information
+    Given I have an account on the site
+    And My account is confirmed
+    When I visit the site root path
+    Then I am presented with a login page
+    When I enter the wrong password
+    And I click the "Log in" button
+    Then I am notified that my password is incorrect
```
Comment (dacur) [inline features/authentication.feature]: your feature files read really nicely!

derek put together a nice guide on cucumber that includes some styling suggestions:
https://github.com/smashingboxes/guides/blob/feature/cucumber-guide/qa/cucumber.md

**Comment by dacur** [comment]
i haven't used cucumber since my apprenticeship, so i'm not :100: on your tests, but they look good to me.  i'm about to start using it on another project and am looking forward to it  :)  a lot of cucumber feedback is stylistic type of stuff and not concrete "this is wrong".

you're doing really well, nice job!  :+1:

**Pull request merged**